### PR TITLE
Pre-compile frontend and add environment config script

### LINKED ISSUE #594: Speed up front container startup by pre-compiling
Compiling is currently done at container startup because we need to take into account the API_URL environment variable (that is compiled into JS).
But this takes about 40 seconds, which is 40 seconds of downtime.

We could simply search and replace the string in the JS but that would break the sourcemap.

Hopefully, there is this little utility: https://www.npmjs.com/package/sourcemap-aware-replace

We could build the JS directly in the container and only search and replace the API_URL string in the files at container startup.

\o/

### PR DESCRIPTION
**Pull request by lukashass**
Closes #594 

- removes replacing environment variables at build time
- add `env-config.js` which gets created using [envsubst](https://github.com/a8m/envsubst) on container startup
- add a webpack dev server middleware that serves `env-config.js` for development

### CODE REVIEW

File: front/Dockerfile [hunk 1]
```diff
@@ -1,23 +1,35 @@
...
-RUN yarn install
+FROM thecodingmachine/nodejs:14-apache
+
+COPY --from=builder --chown=docker:docker /usr/src/front/dist dist
+COPY front/templater.sh .
+
+USER root
+RUN DEBIAN_FRONTEND=noninteractive apt-get update \
+    && apt-get install -y \
+    gettext-base \
```
Comment (anbraten) [inline front/Dockerfile:29]: Is this the package containing `envsubst`?

**Review by anbraten — APPROVED** [review]

**Review by anbraten — APPROVED** [review]

File: front/Dockerfile (hunk 1 above)
Comment (lukashass) [reply front/Dockerfile:29]: Yes

**Comment by moufmouf** [comment]
This looks brilliant. I'll need a bit of time to understand what is going on exactly (especially the Webpack wizardry), but this looks really cool. Thanks!

**Comment by lukashass** [comment]
Great. Let me know when I should clarify anything :wink: 

In the meantime I'll have a look at the build error. Seems like I never tried the build on a clean working directory :see_no_evil:

**Comment by moufmouf** [comment]
Hey @lukashass ,

Build is now passing. There is still an issue with continuous deployment, but the is on my part.

There is still one issue with the unit tests thought (that might prove to be tricky).

We are running unit tests using Jasmine on the "front" container.

(try `yarn run tests`)

Jasmine runs in Node, not in a browser environment.
[So the environment variables are not loaded right now in Jasmine.](https://github.com/thecodingmachine/workadventure/runs/4918521451?check_suite_focus=true)
Could you see if you can load that variables before the tests are run? You will probably need to mock the window object for that.

File: front/src/Enum/EnvironmentVariable.ts [hunk 2]
```diff
@@ -1,30 +1,46 @@
...
-const JITSI_URL: string | undefined = process.env.JITSI_URL === "" ? undefined : process.env.JITSI_URL;
-const JITSI_PRIVATE_MODE: boolean = process.env.JITSI_PRIVATE_MODE == "true";
+declare global {
+    interface Window {
+        env?: Record<string, string>;
+    }
+}
+
+const getEnv = (key: string): string | undefined => {
+    if (!window.env) {
```
Comment (anbraten) [inline front/src/Enum/EnvironmentVariable.ts:8]:
```suggestion
    if (!window || !window.env) {
```

Maybe a simple fix for Jasmine.

Comment (lukashass) [reply front/src/Enum/EnvironmentVariable.ts]: Checking for `!global.window` seems to work. Nice idea!

One thing to note though:
Jasmine will now always use the default values. I might just add a case to return the value from `process.env` if it is set.

**Comment by lukashass** [comment]
I tried `thecodingmachine/workadventure-front:pre-compile-front` in our setup. Seems to work as expected :tada:

**Comment by moufmouf** [comment]
Thanks a lot for this PR! It was really non-trivial and it will make the life of a lot of people easier :heart:

**Pull request merged**